### PR TITLE
connection pooling for duckdb

### PR DESCRIPTION
**Pull request by pjain1**
See details and todos in issue: https://github.com/rilldata/rill-developer/issues/1316

### CODE REVIEW

**Comment by pjain1** [comment]
Modified priority worker test to test for concurrency and priority correctness. Still there are pending things to take care of -
1. Since we are creating DB handles, the way embedded duckdb is behaving is that for each DB handle (created using `sql.Open`) is a separate instance so schemas that are created through one handle are not available on other. If we go with this approach then we will have to limit pool size to 1 for in-memory db.
2. Introduce transactions/callback for handling temp tables used within an api call.
3. More testing to make sure correctness in case of multiple writes because of https://github.com/duckdb/duckdb/issues/77#issuecomment-1104644628

**Comment by begelundmuller** [comment]
Todos:

- [ ] Adapt OLAP tests to reflect different mechanics of priority semaphore
- [ ] Adapt OLAP tests to test with `rill_pool_size > 1`
- [ ] Update timeseries queries to use WithConnection
- [ ] Should we have a separate DB connection we always use for lightweight queries like catalog queries and prepared statements?
- [ ] Should we use `sqlx.Conn` objects instead of `sqlx.DB`? (Also: update terminology in `duckdb/context.go` accordingly)

**Comment by begelundmuller** [comment]
Just a note that I experienced a bug when using this PR with pool size > 1 when doing an initial reconcile against an empty DB. I wonder if some information is not propagated across connections right away. Let's investigate further before merging.

File: runtime/drivers/duckdb/duckdb.go
Comment (begelundmuller) [inline runtime/drivers/duckdb/duckdb.go:36]: This should be `db.SetMaxOpenConns(cfg.PoolSize)` so we don't create more connections than expected. Instead, make the semaphore size:
```go
semSize := cfg.PoolSize - 1
if semSize < 1 {
  semSize = 1
}
```
In the case where `cfg.PoolSize` is 1, it means the meta calls may block a while, but that's fine.

File: runtime/drivers/duckdb/duckdb.go
Comment (begelundmuller) [inline runtime/drivers/duckdb/duckdb.go:51]: Instead of having a `metaConn`, we should use `getConn` in the meta calls. Then it will use the `ctx`'s conn if there is one. 

Since the semaphore size is one less than the pool size, we know there will always be one connection extra in the pool for meta calls.

File: runtime/drivers/duckdb/duckdb.go
Comment (begelundmuller) [inline runtime/drivers/duckdb/duckdb.go:154]: With the DuckDB conn init PR, we can use a built-in connection pool, so we just need to do something like:
```go
conn, err := c.db.Conn()
if err != nil {
  return nil, nil, err
}
return conn, conn.Close, nil
```

File: runtime/drivers/duckdb/migrate.go
Comment (begelundmuller) [inline runtime/drivers/duckdb/migrate.go:10]: Import ordering

File: runtime/drivers/duckdb/migrate.go
Comment (begelundmuller) [inline runtime/drivers/duckdb/migrate.go:88]: `conn` should come after `ctx` in the param list

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Linter errors

File: runtime/drivers/duckdb/duckdb.go
Comment (pjain1) [reply runtime/drivers/duckdb/duckdb.go]: using the built in pool so semSize not needed anymore, for now removed metaConn altogether. Will probably handle that separately

File: runtime/drivers/duckdb/duckdb.go
Comment (pjain1) [reply runtime/drivers/duckdb/duckdb.go]: done

File: runtime/drivers/duckdb/migrate.go
Comment (pjain1) [reply runtime/drivers/duckdb/migrate.go]: done

File: runtime/drivers/duckdb/duckdb.go
Comment (pjain1) [inline runtime/drivers/duckdb/duckdb.go:116]: should we just panic here ?

File: runtime/drivers/duckdb/duckdb.go
Comment (pjain1) [reply runtime/drivers/duckdb/duckdb.go]: when using built in pool, we will have to create and keep `metaConn` explicitly as we don't have control over the pool

File: runtime/drivers/olap.go
Comment (pjain1) [inline runtime/drivers/olap.go:55]: I see this method overrides the `Close()` method on `sqlx.Rows`, is it because in the `Result` struct `sqlx.Rows` parameter does not have a name ? this is new to me.

File: runtime/drivers/duckdb/migrate.go
Comment (pjain1) [inline runtime/drivers/duckdb/migrate.go:29]: why ignore the error from release ? Should we log it in case there is any connection leak ?

File: runtime/queries/table_columns.go
Comment (pjain1) [inline runtime/queries/table_columns.go:76]: should we catch and log error from `rows.Close()` instead of ignoring

File: runtime/drivers/olap.go
Comment (begelundmuller) [reply runtime/drivers/olap.go:55]: Yep :)

File: runtime/drivers/duckdb/migrate.go
Comment (begelundmuller) [reply runtime/drivers/duckdb/migrate.go:29]: I guess we should. Looking in `database/sql`, it looks like the only error it can return is if you try to close the conn twice, so not handling it shouldn't lead to any leaks currently

File: runtime/queries/table_columns.go
Comment (begelundmuller) [reply runtime/queries/table_columns.go:76]: In a perfect world, yeah. We have this many places, we can add a backlog item for it

**Review by begelundmuller — COMMENTED** [review]
I think this is now in a mergable state. I've also done an experience pass, and it seems good. Don't want to merge it until after code freeze, though. Will also do a second review with fresh eyes tomorrow.

File: runtime/drivers/duckdb/migrate.go
Comment (pjain1) [reply runtime/drivers/duckdb/migrate.go:29]: Ok I was looking for a similar thing but didn't knew about it

File: runtime/drivers/olap.go
Comment (pjain1) [reply runtime/drivers/olap.go:55]: Ok I was looking for a similar thing but didn't knew about it

File: runtime/drivers/duckdb/migrate.go
Comment (pjain1) [reply runtime/drivers/duckdb/migrate.go:29]: I guess Go code does not throw any runtime exceptions/errors like Java so it can be ignored and in case of exceptional situations like low memory etc. it should already panic and print in the logs? Only caveat is if in future if API changes and start throwing new error but not a priority to handle now, can be added in the backlog.

Comment (begelundmuller) [reply runtime/drivers/duckdb/migrate.go:29]: Yeah, panics will always propagate. Generally, it's good to handle all errors and only ignore specific named errors, since as you mention, more errors may be returned in future. But for `defer` statements, I often see people pick simpler code. I added a backlog to add good handling for all defer errors.

**Review by begelundmuller — APPROVED** [review]
Second review looks good as well

**Pull request merged**
